tools/pack_glb.py

```diff
@@ -0,0 +1,202 @@
+        for buffer_view in self.description["bufferViews"]:
+            if buffer_view["buffer"] in buffer_offsets:
+                if "byteOffset" not in buffer_view:
+                    buffer_view["byteOffset"] = 0
+                buffer_view["byteOffset"] += buffer_offsets[buffer_view["buffer"]]
+                buffer_view["buffer"] = 0
+
+    def _pack_images(self) -> None:
+        # Rewrite images to reference buffer views inside the internal buffer
+        for image in self.description["images"]:
+            if "uri" in image:
+                uri = urlparse(image["uri"])
+                if uri.scheme != "" or uri.netloc != "":
+                    raise GLTFError(f'support for uri not implemented: "{uri}"')
+                self.source_files.append(Path(uri.path))
+
+                image_size = os.path.getsize(self.source_dir / uri.path)
+
+                image["bufferView"] = len(self.description["bufferViews"])
+                self.description["bufferViews"].append(
+                    {
+                        "buffer": 0,
+                        "byteLength": image_size,
+                        "byteOffset": self.bin_length,
+                    }
+                )
+                del image["uri"]
+
+                self.bin_length += image_size
+
+    def write(self, file: BinaryIO) -> None:
+        """Write a GLB file."""
+        description_json = json.dumps(self.description, separators=(",", ":")).encode(
+            "utf-8"
+        )
+
+        # A GLB file has a JSON chunk and a BIN chunk, each of which need to be aligned to a 4-byte boundary
+        aligned_json_length = align_to_4(len(description_json))
+        json_padding = b" " * (aligned_json_length - len(description_json))
+
+        aligned_bin_length = align_to_4(self.bin_length)
+        bin_padding = b"\x00" * (aligned_bin_length - self.bin_length)
+
+        # write header
+        # magic number 0x46546C67
+        file.write(b"glTF")
+        # version 2
+        file.write((2).to_bytes(4, byteorder="little"))
+        # length (12 byte header + 8 byte chunk header + padded JSON + 8 byte chunk header + padded BIN)
+        length = 12 + 8 + aligned_json_length + 8 + aligned_bin_length
+        file.write(length.to_bytes(4, byteorder="little"))
+
+        # JSON chunk
+        # chunkLength
+        file.write(aligned_json_length.to_bytes(4, byteorder="little"))
+        # chunkType 0x4E4F534A
+        file.write(b"JSON")
+        # chunkData
+        file.write(description_json)
+        file.write(json_padding)
+
+        # BIN chunk
+        # chunkLength
+        file.write(aligned_bin_length.to_bytes(4, byteorder="little"))
+        # chunkType 0x004E4942
+        file.write(b"BIN\x00")
+        # chunkData
+        for source_file_name in self.source_files:
+            with open(self.source_dir / source_file_name, "rb") as source_file:
+                while True:
+                    read = source_file.read(4096)
```

Comment:
why read by 4K? Why not `file.write(source_file.read())` ?
Is it in case the buffer is larger than the ram? Or something I'm missing?
If that's the case, maybe a larger limit would be nice? less syscalls.

Reply:
Yeah, it was intended to avoid using too much RAM. I can just remove the limit for now and add it back if we ever have issues with it